frontend/src/components/pages/RegistrantExperience/index.tsx

```diff
@@ -7,11 +7,109 @@ import Waiver from "./Waiver";
+import {
+  ClickOptionalClause,
+  FillDate,
+  FillName,
+  LoadedWaiver,
+  OptionalClauseResponse,
+  RequiredClauseResponse,
+  WaiverActions,
+  WaiverInterface,
+  WaiverReducerDispatch,
+} from "../../../types/waiverTypes";
+
+const waiverReducer = (
+  waiverInterface: WaiverInterface,
+  action: WaiverReducerDispatch,
+) => {
+  switch (action.type) {
+    case WaiverActions.LOADED_WAIVER: {
+      const optionalClauses: OptionalClauseResponse[] = [];
+      const requiredClauses: RequiredClauseResponse[] = [];
+      const { waiver } = action as LoadedWaiver;
+
+      waiver.clauses.forEach((clause: WaiverClause) => {
+        if (clause.required) requiredClauses.push(clause);
+        else optionalClauses.push({ ...clause, agreed: false });
+      });
+      return {
+        ...waiverInterface,
+        optionalClauses,
+        requiredClauses,
+        waiver,
+        agreedRequiredClauses: false,
+        loadingWaiver: false,
+      };
+    }
+    case WaiverActions.ClICK_REQUIRED_CLAUSE:
+      return {
+        ...waiverInterface,
+        agreedRequiredClauses: !waiverInterface.agreedRequiredClauses,
+      };
+
+    case WaiverActions.CLICK_OPTIONAL_CLAUSE: {
+      if (!waiverInterface.optionalClauses) return waiverInterface;
+      const { optionalClauseId } = action as ClickOptionalClause;
+
+      const newOptionalClauses: OptionalClauseResponse[] = waiverInterface.optionalClauses?.map(
+        (
+          optionalClause: OptionalClauseResponse,
+          index: number,
+        ): OptionalClauseResponse => {
+          if (index === optionalClauseId) {
+            return { ...optionalClause, agreed: !optionalClause.agreed };
+          }
+          return optionalClause;
+        },
+      );
+      return {
+        ...waiverInterface,
+        optionalClauses: newOptionalClauses,
+      };
+    }
+    case WaiverActions.WRITE_NAME: {
+      const { name } = action as FillName;
+      if (name) return { ...waiverInterface, wroteName: true };
+      return { ...waiverInterface, wroteName: false };
+    }
+    case WaiverActions.WRITE_DATE: {
+      const { date } = action as FillDate;
+      if (date) return { ...waiverInterface, wroteDate: true };
+      return { ...waiverInterface, wroteDate: false };
+    }
+    default:
+      return waiverInterface;
+  }
+};
 
 const RegistrantExperiencePage = (): React.ReactElement => {
   const [currentStep, setCurrentStep] = useState<RegistrantExperienceSteps>(
     RegistrantExperienceSteps.PersonalInfoPage,
   );
+  const [waiverInterface, waiverDispatch] = useReducer<
+    Reducer<WaiverInterface, WaiverReducerDispatch>
+  >(waiverReducer, {
+    // TODO: Add support to Waiver and WaiverReducer to get the actual name of the camp being accesses.
+    campName: "Guelph Summer Camp 2022",
+    waiver: undefined,
+    optionalClauses: [],
+    requiredClauses: [],
+    agreedRequiredClauses: false,
+    loadingWaiver: true,
+    wroteDate: false,
+    wroteName: false,
```

Comment:
Can we use these `wroteName`, `wroteDate`, and `agreedRequiredClauses` to evaluate `isWaiverFilled`? Those should be the waiver completion conditions (and then we can delete `sampleWaiverField` and its setter)

Reply:
Yep!